### PR TITLE
scripts: tests: Harness

### PR DESCRIPTION
**Pull request by pkunieck**
Adding more tests for twister harness.

### CODE REVIEW

File: scripts/tests/twister/test_harness.py [hunk 1]
```diff
@@ -34,6 +43,348 @@ def process_logs(harness, logs):
...
+TEST_DATA_11 = [('RunID: 12345', False, False, False, None, True),
+                ('PROJECT EXECUTION SUCCESSFUL', False, False, False, 'passed', False),
+                ('PROJECT EXECUTION SUCCESSFUL', True, False, False, 'failed', False),
+                ('PROJECT EXECUTION FAILED', False, False, False, 'failed', False),
+                ('ZEPHYR FATAL ERROR', False, True, False, None, False),
+                ('GCOV_COVERAGE_DUMP_START', None, None, True, None, False),
+                ('GCOV_COVERAGE_DUMP_END', None, None, False, None, False),]
+@pytest.mark.parametrize(
+    "line, fault, fail_on_fault, cap_cov, exp_stat, exp_id",
+    TEST_DATA_11,
```
Comment (LukaszMrugala) [inline scripts/tests/twister/test_harness.py:55]: We have `TEST_DATA_2`, `TEST_DATA_3`, `TEST_DATA_4`, `TEST_DATA_11`, `TEST_DATA_12`, `TEST_DATA_13` and `TEST_DATA_14` in this file. Those could be reorganised to 1-7.

File: scripts/tests/twister/test_harness.py [hunk 2]
```diff
@@ -34,6 +43,348 @@ def process_logs(harness, logs):
...
+    #Arrange
+    mock_platform = mock.Mock()
+    mock_platform.name = "mock_platform"
+    mock_testsuite = mock.Mock(id = 'id', testcases = [])
+    mock_testsuite.name = "mock_testsuite"
+    instance = TestInstance(testsuite=mock_testsuite, platform=mock_platform, outdir="")
+    instance.testsuite.harness_config = {
+        'robot_test_path': '/path/to/robot/test'
+    }
+    robot_harness = Robot()
```
Comment (LukaszMrugala) [inline scripts/tests/twister/test_harness.py:84]: I would insert a newline between set-ups for different variables if they take up multiple lines. To illustrate my point, this section would be:

```python
    #Arrange
    mock_platform = mock.Mock()
    mock_platform.name = "mock_platform"
    
    mock_testsuite = mock.Mock(id = 'id', testcases = [])
    mock_testsuite.name = "mock_testsuite"
    
    instance = TestInstance(testsuite=mock_testsuite, platform=mock_platform, outdir="")
    instance.testsuite.harness_config = {
        'robot_test_path': '/path/to/robot/test'
    }
    
    robot_harness = Robot()
```

It should help out new readers recognise when a given variable is 'done' and thus help to understand the tests.

File: scripts/tests/twister/test_harness.py [hunk 3]
```diff
@@ -34,6 +43,348 @@ def process_logs(harness, logs):
...
+        else:
+            assert '--device-serial=serial' in command
+            assert '--device-serial-baud=115200' in command
+        assert '--runner=runner' in command
+        assert '--west-flash-extra-args=args' in command
+        assert '--device-id=123' in command
+        assert '--device-product=product' in command
+        assert '--pre-script=pre_script' in command
+        assert '--post-flash-script=post_flash_script' in command
+        assert '--post-script=post_script' in command
```
Comment (LukaszMrugala) [inline scripts/tests/twister/test_harness.py:292]: These could be gathered in some way, like:
```python
    else:
        cmd_list = ['--device-type=hardware',
                    '--runner=runner',
                    '--west-flash-extra-args=args',
                    '--device-id=123',
                    '--device-product=product',
                    '--pre-script=pre_script',
                    '--post-flash-script=post_flash_script',
                    '--post-script=post_script']
                           
        if pty_value == "serial_pty":
            cmd_list += ['--device-serial-pty=serial_pty']
        else:
            cmd_list += ['--device-serial=serial', '--device-serial-baud=115200']
                           
        assert all(cmd in command for cmd in cmd_list)
```

However, the result is worse when it comes to analysing which command is actually missing, so I am only posting this here as a curiosity.

File: scripts/tests/twister/test_harness.py [hunk 4]
```diff
@@ -34,6 +43,348 @@ def process_logs(harness, logs):
...
+    }
+    robot_harness = Robot()
+
+    #Act
+    robot_harness.configure(instance)
+
+    #Assert
+    assert robot_harness.instance == instance
+    assert robot_harness.path == '/path/to/robot/test'
+
```
Comment (LukaszMrugala) [inline scripts/tests/twister/test_harness.py:92]: PEP8 recommends two blank lines between top-level functions. I'd advise that here, as you can tease out where test ends much more easily with that.

**Review by LukaszMrugala — CHANGES_REQUESTED** [review]
Must incorporate 47da4e2e76db4854d02a5b6977c7bd24ee4293bb changes. You should probably either add `harness_config` field to `mock_testsuite` or remove it altogether in the `test_test_handle`. Current implementation fails the aforementioned test wholly.

**Review by LukaszMrugala — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by LukaszMrugala** [comment]
> From a quick glance, it seems this broke CI: https://github.com/zephyrproject-rtos/zephyr/actions/runs/7668713726/job/20901070478

Doesn't seem correct; the pytest-twister-harness tests fail even on 4889c9ca176750341bbaad7c73103608445e4ade, a full day before this merge.

**Comment by henrikbrixandersen** [comment]
Sorry, my mistake. Fix is now in `main`, thanks.

**Pull request merged**